[OFFLOAD] Add plugin with support for Intel oneAPI Level Zero

adurang opened the pull request:
Add a new nextgen plugin that supports GPU devices through the Intel oneAPI Level Zero library.

This PR also adds some new support to the PerThread auxiliary classes that is used by the plugin.

adurang committed:
[OFFLOAD] Add plugin with support for Intel Level Zero

adurang committed:
Merge branch 'main' into l0_plugin

AlexeySachkov commented on offload/plugins-nextgen/level_zero/src/L0Options.cpp:
```diff
@@ -0,0 +1,371 @@
+      StringEnvar("LIBOMPTARGET_LEVEL_ZERO_COMPILATION_OPTIONS", "").get();
+
+  // Explicit Device mode if ONEAPI_DEVICE_SELECTOR is set
+  const StringEnvar DeviceSelectorVar("ONEAPI_DEVICE_SELECTOR", "");
```
It looks weird to me that this plugin recognizes this env variable. The env variable has a wider semantics than just L0 and if we start respecting it, will other plugins (like CUDA & HIP) be updated as well?

adurang replied:
I've removed this for the time being. I agree it probably makes more sense to move it elsewhere.

AlexeySachkov commented on offload/plugins-nextgen/level_zero/src/L0Options.cpp:
```diff
@@ -0,0 +1,371 @@
+
+        std::string_view TopDeviceStr = DeviceSubTuple[0];
+        static const std::array<std::string, 7> DeviceStr = {
+            "host", "cpu", "gpu", "acc", "fpga", "*"};
```
FPGA support has been dropped from intel/llvm (and I assume this env variable is added for seamless transition between the two)

AlexeySachkov commented on offload/plugins-nextgen/level_zero/include/L0Options.h:
```suggestion
  static std::vector<std::string_view> tokenize(const std::string_view &Filter,
```

AlexeySachkov commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1065 @@
+
+        auto maskedId = static_cast<PCIIdTy>(PCIDeviceId & 0xFF00);
+        if (maskedId == Id)
+          return DeviceArchMap[arch].arch; // Exact match or prefix match
```
```suggestion
          return DeviceArchMap[ArchIndex].arch; // Exact match or prefix match
```

AlexeySachkov commented on offload/plugins-nextgen/level_zero/src/L0Options.cpp:
```diff
@@ -0,0 +1,371 @@
+          } else if (!strncmp(DeviceSubTuple[0].data(), "gpu", 3))
+            continue;
+        } else {
+          std::string TDS(TopDeviceStr);
+          if (!isDigits(TDS)) {
+            FAILURE_MESSAGE("error parsing device number: %s",
+                            DeviceSubTuple[0].data());
+          } else {
+            RootD[0] = std::stoi(TDS);
+            DeviceNum = true;
+          }
```
Why do we convert a `string_view` back to `string` in order to check that it only contains digits?

`StringRef` is constructible from `string_view` and has method `getAsInteger` - I suggest we use it instead of custom error checking.

AlexeySachkov commented on offload/plugins-nextgen/level_zero/src/L0Options.cpp:
```diff
@@ -0,0 +1,371 @@
+/// Parse String  and split into tokens of string_views based on the
+/// Delim character.
+std::vector<std::string_view>
+L0OptionsTy::tokenize(const std::string_view &Filter, const std::string &Delim,
```
I would say re-use existing `StringRef::split` and make validation for empty sub-strings at the call site, or have a wrapper around `StringRef::split`, no need to re-implement it.

AlexeySachkov commented on offload/plugins-nextgen/level_zero/include/L0Options.h:
`std::string` is an inefficient common denominator, `StringRef` is a better choice, especially considering that it has [`equals_insensitive`](https://llvm.org/doxygen/classllvm_1_1StringRef.html#ae46058c90a3c703357331a6501b32f1c) method

AlexeySachkov commented on offload/plugins-nextgen/level_zero/include/L0Options.h:
Why not `constexpr std::string_view`?

AlexeySachkov commented on offload/plugins-nextgen/level_zero/include/L0Trace.h:
```diff
@@ -0,0 +1,193 @@
+      DP(__VA_ARGS__);                                                         \
+  } while (0)
+
+#define FATAL_ERROR(Msg)                                                       \
+  do {                                                                         \
+    fprintf(stderr, "%s --> ", DEBUG_PREFIX);                                  \
+    fprintf(stderr, "Error: %s failed (%s) -- exiting...\n", __func__, Msg);   \
+    exit(EXIT_FAILURE);                                                        \
+  } while (0)
```
Why don't re-use existing `FATAL_MESSAGE`, or `FAILURE_MESSAGE`?

adurang replied:
This was old code. I removed it.

AlexeySachkov commented on offload/plugins-nextgen/level_zero/include/L0Trace.h:
```diff
@@ -0,0 +1,193 @@
+  } while (0)
+
+#define WARNING(...)                                                           \
+  do {                                                                         \
+    fprintf(stderr, "%s --> ", DEBUG_PREFIX);                                  \
+    fprintf(stderr, "Warning: " __VA_ARGS__);                                  \
+  } while (0)
```
```suggestion
  DP("Warning: ", __VA_ARGS__)
```

adurang replied:
DP only prints under certain conditions, whereas WARNING prints the message always so these are not equivalent.

adurang commented on offload/plugins-nextgen/level_zero/include/L0Program.h:
```diff
@@ -0,0 +1,135 @@
+  L0ProgramTy() = delete;
+
+  L0ProgramTy(int32_t ImageId, GenericDeviceTy &Device,
+              const __tgt_device_image *Image)
```
do you have a link to where you did that as reference?

adurang committed:
Update offload/CMakeLists.txt

Co-authored-by: Alexey Sachkov <alexey.sachkov@intel.com>

adurang committed:
Update offload/plugins-nextgen/level_zero/CMakeLists.txt

Co-authored-by: Alexey Sachkov <alexey.sachkov@intel.com>

adurang committed:
Update offload/plugins-nextgen/level_zero/include/L0Plugin.h

Co-authored-by: Alexey Sachkov <alexey.sachkov@intel.com>

adurang committed:
Replace pragma once

adurang committed:
Address review comments

adurang committed:
Fix makefile format

adurang committed:
change to StringRef in multiple places

adurang committed:
remove tokenize

adurang committed:
remove unused code

adurang committed:
fix format

adurang committed:
Remove environment variable

adurang committed:
fix getAsInteger conditions

adurang committed:
Merge branch 'main' into l0_plugin

jhuber6 reviewed (commented):
Quick scan, I can't review much beyond nits because I don't know how level zero actually works.

RossBrunton commented on offload/CMakeLists.txt:
```diff
@@ -176,6 +176,19 @@ if(NOT (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64)|(ppc64le)|(aarch64)$"
+        CMAKE_SYSTEM_NAME MATCHES "Linux|Windows"))
+  if("level_zero" IN_LIST LIBOMPTARGET_PLUGINS_TO_BUILD)
+    message(STATUS "Not building Level Zero plugin: it is only supported on "
+	           "Linux/Windows x86_64 or ppc64le hosts")
```
Please don't mix spaces and tabs.

RossBrunton replied:
There's still mixed whitespace. Both in this file and `offload/cmake/Modules/LibomptargetGetDependencies.cmake`.

adurang committed:
Don't use __tgt_device_image

adurang committed:
fix image checking

adurang commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1066 @@
+Expected<InfoTreeNode> L0DeviceTy::obtainInfoImpl() {
+  InfoTreeNode Info;
+  Info.add("Device Number", getDeviceId());
+  Info.add("Device Name", getNameCStr());
```
I'm sorry but I don't quite follow what you want me to do. Can you point me somewhere (example, documentation, ...)?

adurang commented:
> Just tried building this for liboffload - Looks like the platform is returning `OL_PLATFORM_BACKEND_UNKNOWN`, which is hitting the check in initPlugins and causing it to be skipped.

I introduced an issue in the binary checking in a previous change so it might be that. I already push the fix, so do you mind trying again?

adurang committed:
changes from review comments

jhuber6 commented on offload/cmake/Modules/LibomptargetGetDependencies.cmake:
```diff
@@ -89,4 +89,24 @@ if(LIBOMPTARGET_AMDGPU_ARCH)
+
+if(NOT LIBOMPTARGET_DEP_LEVEL_ZERO_INCLUDE_DIRS)
+	set(LIBOMPTARGET_DEP_LEVEL_ZERO_FOUND FALSE)
+  message(STATUS "Could NOT find Level Zero. Missing includes.")
```
I meant all of them

adurang committed:
remove more CMake messages

adurang committed:
format

adurang commented on offload/include/PerThreadTable.h:
```diff
@@ -16,17 +16,93 @@
 #include <mutex>
+#include <type_traits>
+
+template <typename ObjectType> struct PerThread {
```
I introduced the PerThread table to manage some static storage related to Interop. In the plugin, this is used more extensively but the current PerThread didn't support exactly what was needed.
So I reworked it to be a bit more generic and cover the plugin usages  (because I realized I was just rewritting 90% of this again for the plugin).

Other plugins don't have static data AFAIK. But this is just an auxilary data structure so it's not like they couldn't use it if needed.

adurang commented on offload/CMakeLists.txt:
```diff
@@ -176,6 +176,19 @@ if(NOT (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64)|(ppc64le)|(aarch64)$"
     list(REMOVE_ITEM LIBOMPTARGET_PLUGINS_TO_BUILD "cuda")
   endif()
 endif()
+if(NOT (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64)|(AMD64)$" AND
```
All the other plugins are treated in the same way right now (e.g., 3 lines above). They are in the "candidate" list and if some condition is not met they're removed.

I don't disagree with your statement about do it earlier but we should be consistent, so maybe something to do subsequently?

adurang commented on offload/plugins-nextgen/level_zero/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+if(NOT LIBOMPTARGET_DEP_LEVEL_ZERO_FOUND)
+  return()
+endif()
```
This statement is actually unnecessary (a leftover of how it used to work before). If we ever get here it will be built. I'll remove it.

adurang commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,682 @@
+  L0DeviceTLSTy &operator=(const L0DeviceTLSTy &) = delete;
+  L0DeviceTLSTy &operator=(L0DeviceTLSTy &&) = delete;
+
+  auto getCmdList() const { return CmdList; }
```
They could right now, but I had to change many times those "simple member access" to something more complex and this makes that easier imo.

Can we agree to disagree on this one? :)

adurang committed:
remove unnecessary check

adurang commented on offload/plugins-nextgen/level_zero/include/L0Options.h:
```diff
@@ -0,0 +1,157 @@
+/// Specialization constants used for a module compilation.
+class SpecConstantsTy {
+  std::vector<uint32_t> ConstantIds;
+  std::vector<const void *> ConstantValues;
```
They need to be passed as packed-array to L0 so they cannot be unique_ptr unfortunately.

But I'd like to implement your bump pointer suggestion, is there an existing abstraction for that to use?

jhuber6 replied:
What about `unique_ptr` prevents us from having a packed array?
```c++
std::unique_ptr<int[]> arr(new int[N]());
```

jhuber6 replied:
Check the `BumpPtrAllocator Allocator` inside the other runtimes. This is a bump pointer which means the memory never gets freed until the whole class gets freed, but it's perfectly fine for small amounts of data.

pbalcer commented on offload/plugins-nextgen/level_zero/src/L0DynWrapper.cpp:
```diff
@@ -0,0 +1,134 @@
+#define DEBUG_PREFIX "TARGET " GETNAME(TARGET_NAME) " RTL"
+#endif
+
+static bool loadLevelZero() {
```
level-zero supports static linking, which will link in a thin shim layer into the application, and then internally dlopen the actual loader/driver library. Using that mechanism would make this whole dynamic wrapper redundant.

RossBrunton commented:
@adurang I'm now getting compile errors because apparently the L0 plugin is using `ompt_device_t` . Even if it did compile, I think the issue is still present.

`OL_PLATFORM_BACKEND_LEVEL_ZERO` needs adding to `ol_platform_backend_t` and handled in all the places that it is used.

However, thinking about it, I think it might make sense to add liboffload support in another MR and keep this one focusing on libopenmp.

adurang committed:
Merge branch 'main' into l0_plugin

adurang committed:
Add level_zero to liboffload platforms

adurang committed:
fix different issues

adurang committed:
Fix ELF bits from #159623

adurang committed:
format & add more Deviceinfo tags

adurang committed:
tidy up Makefile

adurang committed:
format & renaming

adurang committed:
Remove ompt_device_t reference

adurang commented:
> How well does this pass the unit tests?

This is the current status:

```
Total Discovered Tests: 393
  Unsupported      : 190 (48.35%)
  Passed           :  31 (7.89%)
  Expectedly Failed:  14 (3.56%)
  Failed           : 158 (40.20%)
```

But this is to be expected given than most other pieces for spirv are not yet complete (e.g., frontend) or not there at all (e.g., device library).

jhuber6 commented:
> > How well does this pass the unit tests?
> 
> This is the current status:
> 
> ```
> Total Discovered Tests: 393
>   Unsupported      : 190 (48.35%)
>   Passed           :  31 (7.89%)
>   Expectedly Failed:  14 (3.56%)
>   Failed           : 158 (40.20%)
> ```
> 
> But this is to be expected given than most other pieces for spirv are not yet complete (e.g., frontend) or not there at all (e.g., device library).

I wish someone would implement ballots and stuff for SPIR-V. Then we should be able to port the runtime relatively easily. I don't know if you know anyone who works on the SPIR-V backend but it would be greatly appreciated.

adurang commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,680 @@
+  /// Enqueue copy command
+  int32_t enqueueMemCopy(void *Dst, const void *Src, size_t Size,
+                         __tgt_async_info *AsyncInfo = nullptr,
+                         bool Locked = false, bool UseCopyEngine = true);
```
It doesn't seem to be used anymore. Let me check if we can remove it.

adurang replied:
I removed it

adurang commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1081 @@
+Error L0DeviceTy::dataSubmitImpl(void *TgtPtr, const void *HstPtr, int64_t Size,
+                                 AsyncInfoWrapperTy &AsyncInfoWrapper) {
+  int32_t RC = submitData(TgtPtr, HstPtr, Size, AsyncInfoWrapper);
+  return Plugin::check(RC, "Error in dataSubmitImpl %d", RC);
```
Yes, I know. This is a consequence of Level Zero working on integers and most of the code coming from the old plugin system. It's in my TODO list to have more consistent Error management across the plugin, but it's something I was hoping it can also be delayed for a subsequent PR.

kevinsala replied:
It would be preferable if this is fixed in this PR. Otherwise it will be more difficult to change in the future. I believe it will reduce error handling code in many places, and some extra functions (e.g., `submitData`) could be removed then.

adurang committed:
address reviews

adurang committed:
remove unused argument

adurang commented on offload/plugins-nextgen/level_zero/include/L0Memory.h:
```diff
@@ -0,0 +1,577 @@
+  }
+
+  /// Return the next available buffer
+  void *getNext() {
```
Each allocation done can handle multiple buffers, the offset is to keep track at which point in the allocations the next buffer will be served.

adurang committed:
Merge branch 'main' into l0_plugin

adurang committed:
cleanup

adurang committed:
[OFFLOAD] Enable level zero testing

adurang committed:
address review

adurang committed:
fix & remove dead code

adurang committed:
remove support for target ICVs

adurang committed:
typo & format

adurang committed:
Merge branch 'main' into l0_plugin

adurang committed:
Adjust interfaces after merge

adurang committed:
push some error checking down

jhuber6 commented:
By the way unless this passes all the exists tests it should be disabled by default.

adurang commented:
> By the way unless this passes all the existing tests it should be disabled by default.

There's no testing in this PR.

jhuber6 commented:
> > By the way unless this passes all the existing tests it should be disabled by default.
> 
> There's no testing in this PR.

It probably doesn't enable it as a backend for the OpenMP tests, but I'd imagine it would show up in platform detection for the offloading ones.

adurang commented:
> > > By the way unless this passes all the existing tests it should be disabled by default.
> > 
> > 
> > There's no testing in this PR.
> 
> It probably doesn't enable it as a backend for the OpenMP tests, but I'd imagine it would show up in platform detection for the offloading ones.

Sorry for my ignorance but do you mean the ones in "unittests"?

adurang commented:
> > > By the way unless this passes all the existing tests it should be disabled by default.
> > 
> > 
> > There's no testing in this PR.
> 
> It probably doesn't enable it as a backend for the OpenMP tests, but I'd imagine it would show up in platform detection for the offloading ones.

I checked this, and they're not run. It would need to be explicitly add it in the CMakeLists.txt of unittests.

jhuber6 commented:
Will this have any testing support and does Intel plan on hosting a bot to test it? Because I'm not aware of anyone who can test these changes locally.

adurang commented:
Are they any other comments, issues?

pbalcer commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,625 @@
+                         CmdQueue, 1, &CmdList, nullptr);
+    INFO(OMP_INFOTYPE_PLUGIN_KERNEL, DeviceId,
+         "Submitted kernel " DPxMOD " to device %s\n", DPxPTR(zeKernel), IdStr);
+    CALL_ZE_RET_FAIL(zeCommandQueueSynchronize, CmdQueue, UINT64_MAX);
```
```suggestion
    if (!IsAsync) CALL_ZE_RET_FAIL(zeCommandQueueSynchronize, CmdQueue, UINT64_MAX);
```
?

pbalcer commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1081 @@
+      // kernel are signaled before the kernel is invoked. We always invoke
+      // synchronization on kernel event to support printf().
+      bool WaitDone = false;
+      for (auto Itr = WaitEvents.rbegin(); Itr != WaitEvents.rend(); Itr++) {
```
Might be better to sync on the queue itself, and then just remove all the events. Assuming all events come from the same queue / command list.

pbalcer commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1081 @@
+    Event = getEvent();
+    CALL_ZE_RET_FAIL(zeCommandListAppendMemoryCopy, CmdList, Dst, Src, Size,
+                     Event, 0, nullptr);
+    CALL_ZE_RET_FAIL(zeEventHostSynchronize, Event, UINT64_MAX);
```
You can sync on command list here.

pbalcer commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1081 @@
+    else
+      NumWaitEvents = 0;
+  }
+  auto CmdList = getImmCopyCmdList();
```
This assumes the immediate copy command list exists? The function below checks `if (useImmForCopy()) {`.

adurang committed:
Update offload/plugins-nextgen/level_zero/include/L0Device.h

Co-authored-by: Nick Sarnie <nick.sarnie@intel.com>

adurang committed:
address review comments

pbalcer reviewed (commented):
L0 lgtm.

adurang committed:
format

adurang committed:
more format :/

adurang committed:
add missed wrapper; remove unused macro

adurang committed:
Remove level_zero from default plugin list

adurang committed:
Merge branch 'main' into l0_plugin

adurang committed:
remove removed plugin interface

adurang committed:
Merge branch 'l0_plugin' into l0_tests

adurang committed:
update spirv target

jhuber6 commented on offload/plugins-nextgen/level_zero/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+  message(STATUS "Building Level Zero NG plugin for dlopened level_zero")
+  if(WIN32)
+    cmake_path(REPLACE_EXTENSION LEVEL_ZERO_LIBRARY_NAME dll)
+  endif()
```
Does the L0 library name change? Why can't we just do `#ifdef _WIN32` in the source.

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Defs.h:
```diff
@@ -0,0 +1,67 @@
+namespace llvm::omp::target::plugin {
+
+/// Default alignmnet for allocation
+constexpr size_t L0DefaultAlignment = 0;
```
The default alignment is... infinity?

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,680 @@
+using OmpInteropTy = omp_interop_val_t *;
+class LevelZeroPluginTy;
+
+// clang-format off
```
Why no formatting? Seems like the kind of case where we should just let the clang formatter do its thing. Theres some other cases but those were old code I think.

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,680 @@
+
+  auto getCmdQueue() const { return CmdQueue; }
+  void setCmdQueue(ze_command_queue_handle_t _CmdQueue) {
+    CmdQueue = _CmdQueue;
```
I'm pretty sure `_[A-Z]` identifiers are reserved to the implementation at all scopes in C/C++. Also I've never been a fan of underscores used like this. These really don't need to be getters and setters anyway, this isn't Java.

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,680 @@
+    CacheProperties.stype = ZE_STRUCTURE_TYPE_DEVICE_CACHE_PROPERTIES;
+    CacheProperties.pNext = nullptr;
+
+    auto Err = internalInit();
```
All errors should be propagated back to the top level unless it's an assertion or there's absolutely no way to continue. Why are we doing this here and not in `iniImpl`?

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Kernel.h:
```diff
@@ -0,0 +1,158 @@
+                        uint32_t *GroupSizesOut,
+                        ze_group_count_t &GroupCountsOut,
+                        bool &AllowCooperativeOut) const {
+    if (!LoopDescPtr && memcmp(&LoopDescInit, &LoopDesc, sizeof(LoopDesc)))
```
Why does this need to be a `memcmp`? Guess it doesn't provide a `==` operator do we can't do `*foo == *bar`?

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Memory.h:
```diff
@@ -0,0 +1,579 @@
+  /// Allocation information maintained in the plugin
+  class MemAllocInfoMapTy {
+    /// Map from allocated pointer to allocation information
+    std::map<void *, MemAllocInfoTy> Map;
```
Do we rely on this being a sorted R&B tree or whatever the STL uses? Otherwise a hash table, or specialized LLVM pointer map, is way faster.

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Memory.h:
```diff
@@ -0,0 +1,579 @@
+  /// Release resources and report statistics if requested
+  ~MemAllocatorTy() {
+    if (L0Context)
+      deinit(); // Release resources
```
What's this do? I prefer not to have this type of stuff in destructors. As far as I'm aware most things have an excplit `deinit` that we call at program or whatever teardown.

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Memory.h:
```diff
@@ -0,0 +1,579 @@
+
+  ~StagingBufferTy() {
+    if (initialized())
+      clear();
```
More failable stuff inside of destructors. This is really bad practice when we disable exceptions.

jhuber6 reviewed (commented):
This is a huge patch, would be nice if you could break up some of it.

kevinsala commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1081 @@
+
+Error L0DeviceTy::dataSubmitImpl(void *TgtPtr, const void *HstPtr, int64_t Size,
+                                 AsyncInfoWrapperTy &AsyncInfoWrapper) {
+  int32_t RC = submitData(TgtPtr, HstPtr, Size, AsyncInfoWrapper);
```
As mentioned in the previous comment, it's preferable to fix this in this PR.

kevinsala commented on offload/plugins-nextgen/level_zero/include/L0Plugin.h:
```diff
@@ -0,0 +1,138 @@
+
+  /// L0 device used by each OpenMP device
+  using DeviceContainerTy = llvm::SmallVector<L0DeviceTy *>;
+  DeviceContainerTy L0Devices;
```
Same here. There is a SmallVector of devices in `GenericPluginTy`.

adurang replied:
Yes, but again this was mostly required because subdevices. I think I can move this for the most part to the correct place.

kevinsala commented on offload/plugins-nextgen/level_zero/src/L0Plugin.cpp:
```diff
@@ -0,0 +1,279 @@
+  DP("List of devices (DeviceID[.SubID[.CCSID]])\n");
+  for (auto &l0Device : L0Devices) {
+    DP("-- %s\n", l0Device->getZeIdCStr());
+    (void)l0Device; // silence warning
```
What is it doing?

adurang replied:
It prints (under debug) the low-level identifiers of the devices that were detected.

kevinsala commented on offload/plugins-nextgen/level_zero/src/L0Plugin.cpp:
```diff
@@ -0,0 +1,279 @@
+  auto addDevice = [&DevicesToAdd](auto &zeDevice, auto *Driver, int32_t RootId,
+                                   int32_t SubId = -1, int32_t CCSId = -1) {
+    DevicesToAdd.push_back({{zeDevice, RootId, SubId, CCSId}, Driver});
+  };
```
Is this lambda actually useful?

adurang committed:
address some reviews

adurang committed:
[OFFLOAD] Add support for indexed per-thread containers

Split from #158900 it adds a PerThreadContainer that can use STL-like
indexed containers based on a slightly refactored PerThreadTable.

adurang commented on offload/plugins-nextgen/level_zero/include/L0Defs.h:
```diff
@@ -0,0 +1,67 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef OPENMP_LIBOMPTARGET_PLUGINS_NEXTGEN_LEVEL_ZERO_L0DEFS_H
```
Besides the Plugin::check (which should go away soner or later) all others are not really utils but constants definitions.

adurang committed:
address more reviews

adurang committed:
address more comments

adurang commented on offload/plugins-nextgen/level_zero/include/L0Plugin.h:
```diff
@@ -0,0 +1,138 @@
+
+  auto getDevicesRange() {
+    return DevicesRangeTy(L0Devices.begin(), L0Devices.end());
+  }
```
Ok, do you prefer a separate PR to add this to the common interface?

adurang committed:
change some containers

adurang committed:
more container changes

adurang committed:
move unsafe code out of destructors

adurang committed:
Remove legacy error checking Pt 1

adurang committed:
Remove legacy error checking Pt 2

adurang committed:
Remove legacy error checking Pt 3

adurang committed:
Remove legacy error checking Pt 4

adurang committed:
Add more error checking

adurang commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,680 @@
+        return RC;
+    }
+    return OFFLOAD_SUCCESS;
+  }
```
Ok, I changed all the legacy code so it uses llvm::Error now.

adurang committed:
fix format

adurang committed:
Move DeviceRange to PluginInterface and remove internal Device list

adurang committed:
refactor MemPool options field

adurang committed:
change auto returns

adurang commented:
@jhuber6 @kevinsala I believe I addressed all your comments and questions.

adurang committed:
Remove link copy queues

jhuber6 committed:
Merge branch 'main' into extend_ptt

adurang committed:
Change container

adurang committed:
Add deinit method

adurang committed:
renaming variables, remove some locks

adurang committed:
format

adurang committed:
more renaming

adurang committed:
forgot one mutex

adurang committed:
missed one container

adurang committed:
debug build fixes

adurang committed:
remove auto

adurang committed:
missed one

adurang committed:
Some renaming

adurang committed:
Add asserts

adurang committed:
fixes and cleanup

adurang committed:
format :/

adurang committed:
fixes

kevinsala commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,631 @@
+       "Executing a kernel " DPxMOD "...\n", DPxPTR(zeKernel));
+
+  auto &Plugin = l0Device.getPlugin();
+  auto &Device = Plugin.getDeviceFromId(DeviceId);
```
You already have the device in the parameters.

kevinsala commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,631 @@
+namespace llvm::omp::target::plugin {
+
+Error L0KernelTy::launchImpl(GenericDeviceTy &GenericDevice,
+                             uint32_t NumThreads[3], uint32_t NumBlocks[3],
```
I'm concerned about this. What's the reason your plugin ignores the values for threads and blocks computed in the common interface?

adurang replied:
We actually only support Bare mode for now (until we have more tool-chain support) so those values are the same. But I agree that we should be looking at the parameters to make the path forwar easier so I'd made the change.

kevinsala commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,631 @@
+      }
+      GRPCounts[DistributeDim] = DistributeTripCount;
+    }
+    AllowCooperative = false;
```
Is this always disabled?

kevinsala replied:
If so, please remove the logic depending on this boolean.

kevinsala commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,631 @@
+
+  // We need to get a non-const version of the Properties structure in order to
+  // use its lock and be able to cache the group params and indirect flags
+  auto &KernelPR = const_cast<KernelPropertiesTy &>(getProperties());
```
In this case, I would set the properties as `mutable` in the kernel struct. GetProperties can return a non-const reference keeping the function member as const. And then avoid the const cast.

adurang replied:
Ah, I see. Ok, did that.

kevinsala commented on offload/plugins-nextgen/level_zero/include/L0Trace.h:
```diff
@@ -0,0 +1,194 @@
+      fprintf(stderr, "Error: %s:%s failed with %s\n", __func__, #Fn,          \
+              err_str);                                                        \
+    }                                                                          \
+  } while (0)
```
Not used.

adurang committed:
fix incorrect memory logging and MemPool small refactor

adurang committed:
Refactor code from comments

adurang committed:
Restructure kernel launching code

adurang committed:
Remove unused code

adurang committed:
Use correct thread information

adurang commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,631 @@
+                             AsyncInfoWrapperTy &AsyncInfoWrapper) const {
+
+  auto &l0Device = L0DeviceTy::makeL0Device(GenericDevice);
+  return runTargetTeamRegion(l0Device, KernelArgs, std::move(LaunchParams),
```
See above :)

adurang commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,631 @@
+       GroupCounts.groupCountX, GroupCounts.groupCountY,
+       GroupCounts.groupCountZ);
+  for (int32_t I = 0; I < NumArgs; I++) {
+    {
```
This is to simplify some issues in the code downstream...

adurang commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,631 @@
+  return Plugin::success();
+}
+
+Error L0KernelTy::runTargetTeamRegion(L0DeviceTy &l0Device,
```
I split it in several pieces so I hope is looking better now.

adurang commented:
Thanks @kevinsala. I think I addressed all your comments.

adurang committed:
format

adurang committed:
Fix asserts to not be undefined

adurang committed:
Fix env var parsing issue

adurang committed:
Merge branch 'extend_ptt' into l0_plugin

adurang committed:
Merge branch 'main' into l0_plugin

adurang committed:
add isAccessiblePtrImpl to L0Device

adurang committed:
fixes

adurang committed:
format

adurang committed:
fix

adurang committed:
make kernel properties mutable; remove unused field

adurang committed:
Implement a few additional services

adurang committed:
minor change

adurang committed:
minor refactoring

adurang committed:
Refactor container concepts checks

adurang committed:
remove clearAll

adurang committed:
Simplify PerThreadData

adurang committed:
Merge branch 'extend_ptt' into l0_plugin

adurang committed:
fix

adurang committed:
Address comments

adurang committed:
fix comments

adurang committed:
Merge branch 'extend_ptt' into l0_plugin

adurang committed:
Merge branch 'main' into l0_plugin

adurang committed:
adjust to removal of memory pool

kevinsala commented on offload/plugins-nextgen/level_zero/include/TLS.h:
```diff
@@ -0,0 +1,82 @@
+  bool releaseAsyncQueue(AsyncQueueTy *queue) {
+    if (queue >= &AsyncQueues[0] && queue < &AsyncQueues[PerThreadQueues]) {
+      // it's a local queue
+      queue->InUse = false;
```
I think this is already done in `L0Plugin::releaseAsyncQueue()`

kevinsala commented on offload/plugins-nextgen/common/include/PluginInterface.h:
```diff
@@ -1558,6 +1558,23 @@ struct GenericPluginTy {
   /// object and return immediately.
   int32_t async_barrier(omp_interop_val_t *Interop);
 
+  struct DevicesRangeTy {
```
Please add a bit of documentation on top of the struct declaration.

kevinsala commented on offload/plugins-nextgen/common/include/PluginInterface.h:
```diff
@@ -1291,7 +1291,7 @@ struct GenericPluginTy {
 
   /// Get the reference to the device with a certain device id.
-  GenericDeviceTy &getDevice(int32_t DeviceId) {
+  GenericDeviceTy &getDevice(int32_t DeviceId) const {
```
I remember commenting about this several weeks ago. But I was referring to another solution. I think this change is dangerous. I was referring to have one `getDevice` with const and another without const (even if we duplicate the body code):

```cpp
const GenericDeviceTy &getDevice(int32_t DeviceId) const { ... };
GenericDeviceTy &getDevice(int32_t DeviceId) { ... };
```

kevinsala commented on offload/include/omptarget.h:
```diff
@@ -101,6 +101,7 @@ enum TargetAllocTy : int32_t {
   TARGET_ALLOC_HOST,
   TARGET_ALLOC_SHARED,
   TARGET_ALLOC_DEFAULT,
+  TARGET_ALLOC_LAST = TARGET_ALLOC_DEFAULT
```
Why is this change?

kevinsala commented on offload/liboffload/src/OffloadImpl.cpp:
```diff
@@ -263,6 +263,8 @@ constexpr ol_platform_backend_t pluginNameToBackend(StringRef Name) {
   } else if (Name == "cuda") {
     return OL_PLATFORM_BACKEND_CUDA;
+  } else if (Name == "level_zero") {
+    return OL_PLATFORM_BACKEND_LEVEL_ZERO;
```
I would like to have these liboffload changes in a separate PR. We can merge it after this plugin PR.

adurang commented on offload/plugins-nextgen/level_zero/include/TLS.h:
```diff
@@ -0,0 +1,82 @@
+};
+
+struct L0ThreadTblTy : public PerThread<L0ThreadTLSTy> {
+  void clear() {
```
I think this was needed at some point but not anymore. Cleaned up. Thanks.

adurang commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,472 @@
+                                  uint32_t NumThreads[3],
+                                  uint32_t NumBlocks[3]) const {
+
+  if (KernelEnvironment.Configuration.ExecMode != OMP_TGT_EXEC_MODE_BARE) {
```
It wasn't but I made some changes to make some tests pass. I still wouldn't say they're fully supported but I'm trying not to keep making this PR a moving target...

adurang committed:
cleanup L0ThreadTLS

adurang committed:
use std::bitset to track AsyncQueues

adurang committed:
address comments to changes in PluginInterface

adurang committed:
change usages of *INT_MAX to use numeric_limits

adurang commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,472 @@
+                        CmdQueue, 1, &CmdList, nullptr);
+  INFO(OMP_INFOTYPE_PLUGIN_KERNEL, DeviceId,
+       "Submitted kernel " DPxMOD " to device %s\n", DPxPTR(zeKernel), IdStr);
+  CALL_ZE_RET_ERROR(zeCommandQueueSynchronize, CmdQueue, UINT64_MAX);
```
I adjusted these and all other *_MAX usages. Thanks.

kevinsala replied:
Can you make a constant somewhere and reuse it?

adurang committed:
remove level_zero from pluginNameToBackend

adurang committed:
remove level_zero from liboffload

adurang commented:
Thanks @kevinsala. I think I addressed all your comments.

kevinsala reviewed (approved):
LGTM in general. My review did not focus on Level Zero API specifics, as I am not very familiar with the API. The PR now follows similar structure and style than the rest of plugins, and it seems it re-uses the common plugin interface as expected.

Please wait for more reviewers to approve the PR.

Thanks @adurang for all the changes!

sarnex commented on offload/cmake/Modules/LibomptargetGetDependencies.cmake:
```diff
@@ -89,4 +99,27 @@ if(LIBOMPTARGET_AMDGPU_ARCH)
+  set(LIBOMPTARGET_DEP_LEVEL_ZERO_FOUND FALSE)
+else()
+  set(LIBOMPTARGET_DEP_LEVEL_ZERO_FOUND TRUE)
+  find_library(LIBOMPTARGET_DEP_LEVEL_ZERO_LIBRARY NAMES ze_loader)
```
Do we need to do something if `find_library` fails? If it fails a variable `LIBOMPTARGET_DEP_LEVEL_ZERO_LIBRARY-NOTFOUND` will be defined so we can check that

sarnex commented on offload/plugins-nextgen/level_zero/include/AsyncQueue.h:
```diff
@@ -0,0 +1,52 @@
+
+#include "L0Memory.h"
+
+namespace llvm {
```
nit: i think we can do 
`namespace llvm::omp::target::plugin {` but not a big deal but we already do the same in `L0Context.h`

sarnex commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,643 @@
+  uint32_t getMainCopyEngine() const { return CopyOrdinal.first; }
+
+  bool deviceRequiresImmCmdList() const {
+    return isDeviceIPorNewer(0x05004000);
```
nit: maybe we use a `constexpr` local variable to explain what the magic number argument is?

sarnex commented on offload/plugins-nextgen/level_zero/include/L0Kernel.h:
```diff
@@ -0,0 +1,150 @@
+public:
+  /// Create a L0 kernel with a name and an execution mode.
+  L0KernelTy(const char *Name) : GenericKernelTy(Name), zeKernel(nullptr) {}
+  ~L0KernelTy() {}
```
nit: can do `= default`

sarnex commented on offload/plugins-nextgen/level_zero/include/L0Memory.h:
```diff
@@ -0,0 +1,605 @@
+  MemAllocatorTy(MemAllocatorTy &&) = delete;
+  MemAllocatorTy &operator=(const MemAllocatorTy &) = delete;
+  MemAllocatorTy &operator=(const MemAllocatorTy &&) = delete;
+  ~MemAllocatorTy() {}
```
nit: `= default`

sarnex commented on offload/plugins-nextgen/level_zero/include/L0Memory.h:
```diff
@@ -0,0 +1,605 @@
+  StagingBufferTy &operator=(const StagingBufferTy &) = delete;
+  StagingBufferTy &operator=(const StagingBufferTy &&) = delete;
+
+  ~StagingBufferTy() {}
```
nit: take a guess

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1103 @@
+  IndirectAccessFlags = Flags;
+
+  // Get the UUID
+  std::string uid = "";
```
nit: no need for `="";`

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1103 @@
+
+  auto &Plugin = getPlugin();
+
+  AsyncQueueTy *AsyncQueue = (AsyncQueueTy *)AsyncInfo.Queue;
```
nit: can we avoid c cast?

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1103 @@
+      TgtPtrType == ZE_MEMORY_TYPE_SHARED) {
+    bool CopyNow = true;
+    if (IsAsync) {
+      if (AsyncQueue->KernelEvent) {
```
nit: could we make this a single if like `if(IsAsync && AsyncQueue->KernelEvent)`?

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1103 @@
+  // non-ordered unless is targetsync
+  return interop_spec_t{
+      tgt_fr_level_zero,
+      {InteropType == kmp_interop_type_targetsync ? true : false /*inorder*/,
```
nit: i think we can remove the `? true : false `

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1103 @@
+Expected<OmpInteropTy> L0DeviceTy::createInterop(int32_t InteropContext,
+                                                 interop_spec_t &InteropSpec) {
+  auto Ret =
+      new omp_interop_val_t(DeviceId, (kmp_interop_type_t)InteropContext);
```
nit: c-cast again sry

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,474 @@
+    DP("Max team size execceds current maximum %zu. Adjusted\n",
+       MaxThreadLimit);
+  }
+  {
```
nit: if theres not a reason to scope this block with `{}` i recommend removing it, i don't see a reason to do it

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Kernel.cpp:
```diff
@@ -0,0 +1,474 @@
+
+  // Set kernel arguments
+  for (int32_t I = 0; I < NumArgs; I++) {
+    {
```
nit: the scoping `{}` seems unnecessary to me

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Memory.cpp:
```diff
@@ -0,0 +1,720 @@
+    return reinterpret_cast<void *>(Base + I * ChunkSize);
+  }
+  // Should not reach here.
+  assert(0 && "Inconsistent memory pool state");
```
nit: `0`->`false`

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Memory.cpp:
```diff
@@ -0,0 +1,720 @@
+                         "Cannot find base address of " DPxMOD "\n",
+                         DPxPTR(Ptr));
+  }
+  log(0, Info.AllocSize, Info.Kind);
```
nit: maybe an inline comment of what the `0` is?

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Memory.cpp:
```diff
@@ -0,0 +1,720 @@
+    HostDesc.pNext = &RelaxedDesc;
+  }
+
+  auto zeDevice = Device ? Device->getZeDevice() : 0;
```
nit: can we do `0`->`nullptr` since its a handle?

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Plugin.cpp:
```diff
@@ -0,0 +1,249 @@
+
+  struct RootInfoTy {
+    uint32_t OrderId;
+    ze_device_handle_t zeDevice;
```
nit: capital Z?

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Program.cpp:
```diff
@@ -0,0 +1,552 @@
+    return Plugin::success();
+  }
+
+  if (Modules.empty()) {
```
nit: single stmt `{}`

sarnex commented on offload/plugins-nextgen/level_zero/src/L0Program.cpp:
```diff
@@ -0,0 +1,552 @@
+  // don't know which module contains a specific kernel with a name.
+  for (auto Module : Modules) {
+    uint32_t Count = 0;
+    CALL_ZE_RET_ERROR(zeModuleGetKernelNames, Module, &Count, nullptr);
```
nit: inline comment on what `nullptr` is?

sarnex reviewed (commented):
a long list of c++ style nits

adurang committed:
fix typo

adurang committed:
Make a constant for timeout expression

adurang committed:
address comments

adurang committed:
address comments

adurang commented:
Thanks @sarnex. I think I addressed all your comments.

adurang committed:
format

sarnex reviewed (approved):
thanks so much for addressing everything! just a couple more nits but this lgtm!

adurang committed:
address comments

adurang commented:
As there are two approvals for this PR, I plan to land these changes by Dec 18th. If anyone has feedback you’d like addressed before I land the changes, please let me know, otherwise we can handle further issues in post-commit. Thanks!
(Remember this plugin is not enabled by default)

jhuber6 commented on offload/plugins-nextgen/level_zero/include/AsyncQueue.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace llvm::omp::target::plugin {
+
+/// Abstract queue that supports asynchronous command submission
```
All comments should use proper punctuation in LLVM, here and elsewhere

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,644 @@
+  void clear() {
+    // destroy all lists and queues
+    if (CmdList)
+      CALL_ZE_EXIT_FAIL(zeCommandListDestroy, CmdList);
```
We're not the best at this in the rest of the library, but we *really* should not have this type of failure anywhere in a real runtime. This should return an error and if it's a fatal error that should be handled at a higher level.

jhuber6 commented on offload/plugins-nextgen/level_zero/include/L0Device.h:
```diff
@@ -0,0 +1,644 @@
+  ze_command_list_handle_t getImmCopyCmdList() const { return ImmCopyCmdList; }
+  void setImmCopyCmdList(ze_command_list_handle_t ImmCopyCmdListIn) {
+    ImmCopyCmdList = ImmCopyCmdListIn;
+  }
```
I still don't like this useless cargo cult OOP pattern, either just return a reference or keep it as a standard member. This is only useful when getting and setting does non-trivial work and/or validation.

jhuber6 commented on offload/plugins-nextgen/level_zero/src/L0Device.cpp:
```diff
@@ -0,0 +1,1095 @@
+
+DeviceArchTy L0DeviceTy::computeArch() const {
+  const auto PCIDeviceId = getPCIId();
+  if (PCIDeviceId != 0) {
```
Prefer early exit over nesting

jhuber6 reviewed (commented):
Some comments, I really want to minimize the use of macros and a proper runtime should never, ever `if(bad) die()`.

adurang committed:
Address comments

adurang committed:
remove some macros

adurang committed:
remove unusued macro

adurang committed:
remove str macro

adurang committed:
Fix comments punctuation Pt.1

adurang committed:
Fix comments punctuation Pt.2

adurang committed:
Simplify getDevicesRange

adurang committed:
space

adurang committed:
Fix level zero library name

jhuber6 reviewed (approved):
Seems fine, but I'm sure there's something that's slipped through the cracks while reviewing 6000 lines of a runtime I don't know. Next steps will be to get a bot running green and then we can start discussions about enabling this by default
